File: 10. Stack and Queue/4. Implementation Problems/Rotting-Oranges.py
# Rotting Oranges (Leetcode)
# Leetcode Link : https://leetcode.com/problems/rotting-oranges/description/

# Time Complexity : O(row * col)
# Space Complexity : O(row * col)

from collections import deque

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        row = len(grid)
        col = len(grid[0])
        
        # Initialize a deque for BFS and a visited matrix
        que = deque()
        visited = [[0] * col for _ in range(row)]
        countFresh = 0  # To count the fresh oranges

        for i in range(row):
            for j in range(col):
                if grid[i][j] == 2:
                    que.append([i, j, 0])  # Append rotten orange's position and time 0
                    visited[i][j] = 2  # Mark the cell as visited
                else:
                    visited[i][j] = 0
                
                if grid[i][j] == 1:
                    # Increment fresh orange counter
                    countFresh += 1 

        time = 0  # To track the time taken to rot all oranges
        # Directions for the 4 possible adjacent cells (up, right, down, left)
        delRow = [-1, 0, 1, 0]
        delCol = [0, 1, 0, -1]
        count = 0  # To count the rotted fresh oranges

        # BFS to rot all adjacent fresh oranges
        while que:
            r = que[0][0]
            c = que[0][1]
            t = que[0][2]
            que.popleft()  # Remove the current cell from the queue
            time = max(time, t)  # Update the time with the maximum time value encountered

            for i in range(4):
                nrow = r + delRow[i]
                ncol = c + delCol[i]
                # If the adjacent cell is within bounds, is a fresh orange and hasn't been visited
                if 0 <= nrow < row and 0 <= ncol < col and visited[nrow][ncol] == 0 and grid[nrow][ncol] == 1:
                    # Append the new rotten orange to the queue with incremented time
                    que.append([nrow, ncol, t + 1])
                    # Mark the cell as visited
                    visited[nrow][ncol] = 2
                    # Increment the counter for rotted fresh oranges
                    count += 1
            
        # If the number of fresh oranges rotted doesn't match the initial fresh oranges count, return -1
        if countFresh != count:
            return -1
        
        return time

File: 19. Graphs/2. BFS & DFS/Rotting-Oranges.cpp
// Rotting Oranges (Leetcode)
// Leetcode Link : https://leetcode.com/problems/rotting-oranges/description/

// Time Complexity : O(m * n)
// Space Complexity : O(m * n)

class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        queue<vector<int>> q;
        int minutes = 0;

        // Step 1: Push all initially rotten oranges into the queue
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(grid[i][j] == 2){
                    q.push({grid[i][j], i, j});
                }
            }
        }

        // Step 2: Perform BFS to spread the rot to neighboring fresh oranges
        while(!q.empty()){
            int num = q.size();

            for(int i = 0; i < num; i++){
                vector<int> orange = q.front();
                q.pop();

                int row = orange[1];
                int col = orange[2];

                int delRow[] = {-1, 0, 1, 0};
                int delCol[] = {0, 1, 0, -1};

                // Step 3: Check all 4 neighboring cells for fresh oranges
                for(int k = 0; k < 4; k++){
                    int nRow = row + delRow[k];
                    int nCol = col + delCol[k];

                    if(nRow >= 0 && nRow < m && nCol >= 0 && nCol < n && grid[nRow][nCol] == 1){
                        grid[nRow][nCol] = 2;
                        q.push({grid[nRow][nCol], nRow, nCol});
                    }
                }
            }

            if(!q.empty()) minutes++;
        }

        // Step 4: Check if there are any fresh oranges left
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(grid[i][j] == 1) return -1;
            }
        }

        // Step 5: Return the time taken for all oranges to rot
        return minutes;
    }
};